docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+
+* how is this going to interoperate with Verona and the target language
+* how the remote procedure calls will be dispatched and the arguments serialised
+* how the runtime will separate the memory from the main process and sanitise
+  the execution (system calls, file descriptors, ioctl, etc)
+
+## Overall architecture
+
+Each external language needs explicit support in the Verona compiler to be
+parsed at compile time and generate the correct stubs for the interoperability
+layer to correctly implement the remote procedure call (RPC).
+
+Each execution model (child process, web assembly, etc) will need explicit
+support for deploying the functionality at run time, hopefully not individually
+bound to the source language (N-to-M).
+
+But the overall architecture will be the same.
+
+### Parser
+
+1. Upon finding a file include that has foreign origin, the appropriate foreign
+   language parser (clang for C++) will be asked to parse the file and generate
+   its own AST.
+2. A Verona module will be created, with all foreign builtin types.
+3. When the Verona compiler asks for a particular type or function in the foreign
+   module, the language parser will convert the signature of its own AST into
+   Verona AST and add that node in its own module, then pass that to the Verona
+   AST node.
+
+This allows lazy evaluation of what could be very large number of includes, with
+code that can't be translated into Verona.
+
+The end result is an AST that has foreign nodes as Verona nodes on Verona modules
+without the actual implementation. For all purposes, this is a plain Verona AST
+with calls to functions that have been declared, but not defined.
+
+Definition (and implementation in object code) will be done later by the compiler
+once we know all functions that have survived the initial round of cleanups (ex.
+reachability analysis).
+
+### Compiler
+
+1. For each sandbox region, the compiler generates calls to the runtime to setup
+   the region, including its allocators.
+2. Upon finding a foreign call, the compiler will ask the language driver for
+   the foreign AST node, to implement the actual call.
+3. The driver will generate the conversion between Verona and native types,
+   including structure layout handling per architecture, serialise the
+   arguments and return values into a buffer and call the dispatcher below.
+4. A dispatcher will be created to call those functions with serialised
+   arguments by an index that is associated when they're created.
+5. A bi-directional RPC layer is created between the Verona code and the sandbox,
+   where calls to external functions go one way, and calls to allocation, system
+   calls, etc. go the other way.
+6. The appropriate object files are created and linked with the main executable.
+
+Not all functions will be representable in Verona, not all structure layouts will
+be easily known at the Verona level (special hardware ABI), so errors may still
+occur here.
```

Comment:
Can you give an example of the kind of thing here that might be a problem?  I am expecting that we'll only ever pass primitives / pointers through the translation layer.  If there's a struct with a funky layout, we'll handle it in the same way as any other struct: Generate RPC calls for reading / writing the fields and then seeing if we can inline them in the caller safely later (i.e. if, after lowering them to LLVM IR, we can insert checks for load and store instructions and if there are no calls).

Reply:
This sentence to me implies things that will be correctly parsed but where we generate incorrect code because we don't have enough information to implement it correctly.  I can't think of anything where I'd expect that to happen because we either support a feature or we don't.  I agree that we won't support everything, but I don't think we should be enabling any C/C++ language construct until we are happy that it either works or reports an error.

Reply:
Ah, when I said "error may occur" I meant "report an error". I'll clarify.